File: src/humanize/typing.rs
//! Typing mutation — human-like keystroke rhythm and error handling

use rand::Rng;

use super::config::HumanizationConfig;

/// A single typing event
#[derive(Debug, Clone)]
pub enum TypingEvent {
    /// Type a single character
    Key { ch: char, interval_ms: u32 },
    /// Backspace (correction)
    Backspace { interval_ms: u32 },
    /// Pause mid-text (thinking)
    Pause { duration_ms: u32 },
}

/// Full typing plan for a string
#[derive(Debug, Clone)]
pub struct TypingPlan {
    pub events: Vec<TypingEvent>,
    /// Total time to type the full string (ms)
    pub total_ms: u32,
}

/// Build a humanized typing plan from raw text + config
pub fn build_typing_plan(text: &str, config: &HumanizationConfig) -> TypingPlan {
    let mut events = Vec::new();
    let mut total_ms: u32 = 0;
    let mut rng = rand::thread_rng();
    let pattern = &config.typing;

    // Edge case
    if text.is_empty() {
        return TypingPlan { events, total_ms: 0 };
    }

    for (i, ch) in text.chars().enumerate() {
        // Insert pre-character delay (thinking/reaction time)
        let pre_delay = pattern.speed_variance_percent.min(30) as u32;
        let pre_ms = rng.gen_range(0..=pre_delay);
        if pre_ms > 0 {
            total_ms += pre_ms;
            events.push(TypingEvent::Pause { duration_ms: pre_ms });
        }

        // Occasional mid-text "thinking" pause
        if pattern.pause_chance > 0.0 && rng.gen::<f32>() < pattern.pause_chance {
            let pause = pattern.pause_duration_ms;
            total_ms += pause;
            events.push(TypingEvent::Pause { duration_ms: pause });
        }

        // Determine base interval for this character
        let base_interval = 60000 / pattern.base_wpm.max(1);
        let variance = (base_interval as f32 * pattern.speed_variance_percent as f32 / 100.0) as u32;
        let interval = if variance > 0 {
            rng.gen_range(base_interval.saturating_sub(variance)..=base_interval + variance)
        } else {
            base_interval
        };

        // Special keys: shift for uppercase takes longer, punctuation slightly longer
        let adjusted_interval = if ch.is_uppercase() || ch.is_ascii_punctuation() {
            interval.saturating_add(25)
        } else {
            interval
        };

        // Simulate typo + correction
        if pattern.error_retry_chance > 0.0 && rng.gen::<f32>() < pattern.error_retry_chance {
            // Type wrong character first
            events.push(TypingEvent::Key { ch, interval_ms: adjusted_interval / 2 });
            total_ms += adjusted_interval / 2;

            // Brief pause to "notice" the mistake
            let notice_pause: u32 = rng.gen_range(150..=350);
            total_ms += notice_pause;
            events.push(TypingEvent::Pause { duration_ms: notice_pause });

            // Backspace
            events.push(TypingEvent::Backspace { interval_ms: 80 });
            total_ms += 80;

            // Retype correctly
            events.push(TypingEvent::Key { ch, interval_ms: adjusted_interval });
            total_ms += adjusted_interval;
        } else {
            events.push(TypingEvent::Key { ch, interval_ms: adjusted_interval });
            total_ms += adjusted_interval;
        }

        // End of word: small natural pause
        if ch.is_whitespace() && i < text.len() - 1 {
            let word_pause: u32 = rng.gen_range(30..=80);
            total_ms += word_pause;
            events.push(TypingEvent::Pause { duration_ms: word_pause });
        }
    }

    TypingPlan { events, total_ms }
}

/// Compute single-character interval without building full plan (for inline use)
pub fn char_interval(config: &HumanizationConfig, ch: char) -> u32 {
    let pattern = &config.typing;
    let base = 60000 / pattern.base_wpm.max(1);
    let variance = (base as f32 * pattern.speed_variance_percent as f32 / 100.0) as u32;

    let mut rng = rand::thread_rng();
    let interval = if variance > 0 {
        rng.gen_range(base.saturating_sub(variance)..=base + variance)
    } else {
        base
    };

    if ch.is_uppercase() || ch.is_ascii_punctuation() {
        interval.saturating_add(25)
    } else {
        interval
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn medium_config() -> HumanizationConfig {
        HumanizationConfig::from_level(super::super::config::HumanizationLevel::Medium)
    }

    #[test]
    fn test_empty_text() {
        let plan = build_typing_plan("", &medium_config());
        assert!(plan.events.is_empty());
        assert_eq!(plan.total_ms, 0);
    }

    #[test]
    fn test_hello_plan() {
        let config = medium_config();
        let plan = build_typing_plan("hello", &config);
        assert!(!plan.events.is_empty());
        assert!(plan.total_ms > 0);
        // Should contain at least 5 key events (one per char)
        let key_count = plan.events.iter().filter(|e| matches!(e, TypingEvent::Key { .. })).count();
        assert_eq!(key_count, 5, "should have 5 key events for 'hello'");
    }

    #[test]
    fn test_no_humanization() {
        let config = HumanizationConfig::from_level(super::super::config::HumanizationLevel::None);
        let plan = build_typing_plan("test", &config);
        // With None level, there should be no pauses or corrections
        let pauses: usize = plan.events.iter().filter(|e| matches!(e, TypingEvent::Pause { .. })).count();
        assert_eq!(pauses, 0, "no pauses when humanization is None");
    }
}

File: backend/internal/behavior/humanize/typing.go
package humanize

import "math/rand/v2"

// TypingEventType distinguishes keys, backspaces, and pauses in a typing plan.
type TypingEventType int

const (
	TypingEventKey       TypingEventType = 0
	TypingEventBackspace TypingEventType = 1
	TypingEventPause     TypingEventType = 2
	TypingEventModifier  TypingEventType = 3
	TypingEventIME       TypingEventType = 4
	TypingEventTab       TypingEventType = 5
	TypingEventClipboard TypingEventType = 6
)

// TypingEvent is one step in a typing plan.
type TypingEvent struct {
	Type       TypingEventType
	Ch         rune   // only for Key
	IntervalMs uint32 // for Key and Backspace
	DurationMs uint32 // for Pause
	Action     string // for Modifier/IME/Clipboard
}

// TypingPlan is a complete sequence of typing events for a text string.
type TypingPlan struct {
	Events  []TypingEvent
	TotalMs uint32
}

// BuildTypingPlan builds a complete humanized typing plan for the given text.
// Includes per-character intervals, typos with backspace corrections,
// mid-text thinking pauses, and word-boundary pauses.
func BuildTypingPlan(text string, config *HumanizationConfig) TypingPlan {
	events := make([]TypingEvent, 0, len(text)*2)
	var totalMs uint32

	if len(text) == 0 {
		return TypingPlan{Events: events, TotalMs: 0}
	}

	cfg := &config.Typing
	runes := []rune(text)

	for i, ch := range runes {
		// 1. Pre-character "reaction" jitter
		preDelayMax := min(30, cfg.SpeedVariancePercent)
		if preDelayMax > 0 {
			preMs := randRangeUint32(0, preDelayMax)
			if preMs > 0 {
				totalMs += preMs
				events = append(events, TypingEvent{Type: TypingEventPause, DurationMs: preMs})
			}
		}

		// 2. Occasional mid-text thinking pause
		if cfg.PauseChance > 0 && rand.Float64() < cfg.PauseChance {
			totalMs += cfg.PauseDurationMs
			events = append(events, TypingEvent{Type: TypingEventPause, DurationMs: cfg.PauseDurationMs})
		}

		// 3. Compute base interval for this character
		baseInterval := uint32(60000 / max(1, cfg.BaseWPM))
		variance := uint32(float64(baseInterval) * float64(cfg.SpeedVariancePercent) / 100.0)
		var interval uint32
		if variance > 0 {
			interval = randRangeUint32(baseInterval-variance, baseInterval+variance)
		} else {
			interval = baseInterval
		}

		// 4. Shift/caps/punctuation penalty
		adjustedInterval := interval
		if (ch >= 'A' && ch <= 'Z') || isASCIISymbol(ch) {
			adjustedInterval = interval + 25
		}

		// 5. Typo simulation with retry
		if cfg.ErrorRetryChance > 0 && rand.Float64() < cfg.ErrorRetryChance {
			// Type wrong character (half interval)
			wrongCh := adjacentKey(ch)
			events = append(events, TypingEvent{Type: TypingEventKey, Ch: wrongCh, IntervalMs: adjustedInterval / 2})
			totalMs += adjustedInterval / 2

			// "Notice" pause
			noticePause := randRangeUint32(150, 350)
			totalMs += noticePause
			events = append(events, TypingEvent{Type: TypingEventPause, DurationMs: noticePause})

			// Backspace
			events = append(events, TypingEvent{Type: TypingEventBackspace, IntervalMs: 80})
			totalMs += 80

			// Retype correctly (full interval)
			events = append(events, TypingEvent{Type: TypingEventKey, Ch: ch, IntervalMs: adjustedInterval})
			totalMs += adjustedInterval
		} else {
			// Normal keystroke
			events = append(events, TypingEvent{Type: TypingEventKey, Ch: ch, IntervalMs: adjustedInterval})
			totalMs += adjustedInterval
		}

		// 6. End-of-word pause (after whitespace, unless last char)
		if ch == ' ' && i < len(runes)-1 {
			wordPause := randRangeUint32(30, 80)
			totalMs += wordPause
			events = append(events, TypingEvent{Type: TypingEventPause, DurationMs: wordPause})
		}
	}

	return TypingPlan{Events: events, TotalMs: totalMs}
}

// CharInterval returns the typing interval for a single character (without typo logic).
// Useful for inline estimates when a full plan is not needed.
func CharInterval(config *HumanizationConfig, ch rune) uint32 {
	cfg := &config.Typing
	base := uint32(60000 / max(1, cfg.BaseWPM))
	variance := uint32(float64(base) * float64(cfg.SpeedVariancePercent) / 100.0)
	var interval uint32
	if variance > 0 {
		interval = randRangeUint32(base-variance, base+variance)
	} else {
		interval = base
	}
	if (ch >= 'A' && ch <= 'Z') || isASCIISymbol(ch) {
		interval += 25
	}
	return interval
}

// adjacentKey returns a key "near" the given key on a QWERTY layout
// to simulate a common typo.
func adjacentKey(ch rune) rune {
	neighbors := map[rune]rune{
		'a': 's', 's': 'a', 'd': 's', 'f': 'd', 'g': 'f',
		'h': 'g', 'j': 'h', 'k': 'j', 'l': 'k',
		'q': 'w', 'w': 'q', 'e': 'w', 'r': 'e', 't': 'r',
		'y': 't', 'u': 'y', 'i': 'u', 'o': 'i', 'p': 'o',
		'z': 'x', 'x': 'z', 'c': 'x', 'v': 'c', 'b': 'v',
		'n': 'b', 'm': 'n',
	}
	lower := ch
	if ch >= 'A' && ch <= 'Z' {
		lower = ch + 32
	}
	if n, ok := neighbors[lower]; ok {
		if ch >= 'A' && ch <= 'Z' {
			return n - 32
		}
		return n
	}
	return ch
}

// TypingContext tunes intervals for field-level typing semantics.
type TypingContext struct {
	FieldKind        string
	MinutesElapsed   uint32
	CaptchaCellCount uint32
}

func BuildContextTypingPlan(text string, context TypingContext, config *HumanizationConfig) TypingPlan {
	plan := BuildTypingPlan(text, config)
	multiplier := 1.0 + float64(context.MinutesElapsed)*0.025
	if context.FieldKind == "password" {
		multiplier *= 1.4
	}
	if multiplier != 1.0 {
		for i := range plan.Events {
			if plan.Events[i].IntervalMs > 0 {
				plan.Events[i].IntervalMs = uint32(float64(plan.Events[i].IntervalMs) * multiplier)
			}
			if plan.Events[i].DurationMs > 0 {
				plan.Events[i].DurationMs = uint32(float64(plan.Events[i].DurationMs) * multiplier)
			}
		}
		plan.TotalMs = sumTypingPlanMs(plan.Events)
	}
	if context.FieldKind == "captcha" && context.CaptchaCellCount > 1 {
		plan = insertCaptchaCellPauses(plan, context.CaptchaCellCount)
	}
	return plan
}

func BuildIMEPlan(pinyin string, candidateMoves uint32, config *HumanizationConfig) TypingPlan {
	plan := BuildTypingPlan(pinyin, config)
	for i := uint32(0); i < candidateMoves; i++ {
		plan.Events = append(plan.Events, TypingEvent{Type: TypingEventIME, Action: "candidate_arrow", IntervalMs: randRangeUint32(120, 260)})
	}
	plan.Events = append(plan.Events, TypingEvent{Type: TypingEventIME, Action: "candidate_enter", IntervalMs: randRangeUint32(120, 220)})
	plan.TotalMs = sumTypingPlanMs(plan.Events)
	return plan
}

func BuildModifierKeyPlan(sequence string, config *HumanizationConfig) TypingPlan {
	interval := uint32(90)
	if config != nil {
		interval = max(60, config.Timing.PreActionDelayMs+40)
	}
	return TypingPlan{Events: []TypingEvent{
		{Type: TypingEventModifier, Action: "ctrl_down", IntervalMs: interval},
		{Type: TypingEventModifier, Action: sequence, IntervalMs: randRangeUint32(45, 90)},
		{Type: TypingEventModifier, Action: "ctrl_up", IntervalMs: randRangeUint32(30, 70)},
	}, TotalMs: interval + 90 + 70}
}

func BuildTabSwitchPlan(fields uint32) TypingPlan {
	if fields == 0 {
		return TypingPlan{}
	}
	events := make([]TypingEvent, 0, fields)
	var total uint32
	for i := uint32(0); i < fields; i++ {
		interval := uint32(sampleNormal(500, 200, 700))
		if interval < 120 {
			interval = 120
		}
		events = append(events, TypingEvent{Type: TypingEventTab, Action: "tab", IntervalMs: interval})
		total += interval
	}
	return TypingPlan{Events: events, TotalMs: total}
}

func BuildClipboardPlan(config *HumanizationConfig) TypingPlan {
	plan := TypingPlan{}
	for _, action := range []string{"ctrl+a", "ctrl+c", "ctrl+v"} {
		part := BuildModifierKeyPlan(action, config)
		plan.Events = append(plan.Events, part.Events...)
	}
	plan.TotalMs = sumTypingPlanMs(plan.Events)
	return plan
}

func FingerSpeedRatio(ch rune) float64 {
	switch ch {
	case 'q', 'a', 'z', 'p':
		return 0.40
	case 'w', 's', 'x', 'o', 'l':
		return 0.55
	case 'e', 'd', 'c', 'i', 'k':
		return 0.70
	case 'r', 't', 'f', 'g', 'v', 'b', 'y', 'u', 'h', 'j', 'n', 'm':
		return 0.80
	default:
		return 0.65
	}
}

func sumTypingPlanMs(events []TypingEvent) uint32 {
	var total uint32
	for _, event := range events {
		total += event.IntervalMs + event.DurationMs
	}
	return total
}

func insertCaptchaCellPauses(plan TypingPlan, cells uint32) TypingPlan {
	if cells <= 1 {
		return plan
	}
	keysSeen := uint32(0)
	next := make([]TypingEvent, 0, len(plan.Events)+int(cells))
	for _, event := range plan.Events {
		next = append(next, event)
		if event.Type == TypingEventKey {
			keysSeen++
			if keysSeen < cells {
				next = append(next, TypingEvent{Type: TypingEventPause, DurationMs: randRangeUint32(180, 420)})
			}
		}
	}
	return TypingPlan{Events: next, TotalMs: sumTypingPlanMs(next)}
}
